File: 3rd_part_RO_lab_algorithms_CLI/python/ex6_climb_stairs.py
def climb_stairs(n, memo=None):
    if memo is None:
        memo = {}
    if n == 0:
        return 1
    if n < 0:
        return 0
    if n in memo:
        return memo[n]
    memo[n] = climb_stairs(n - 1, memo) + climb_stairs(n - 2, memo) + climb_stairs(n - 3, memo)
    return memo[n]
print(climb_stairs(4))

File: 3rd_part_RO_lab_algorithms_CLI/c/ex6_climb_stairs.c
#include <stdio.h>
int memo[100] = {0};
int climb_stairs(int n) {
    if (n == 0) return 1;
    if (n < 0) return 0;
    if (memo[n] != 0) return memo[n];
    memo[n] = climb_stairs(n - 1) + climb_stairs(n - 2) + climb_stairs(n - 3);
    return memo[n];
}
int main() {
    printf("%d\n", climb_stairs(4));
    return 0;
}
